### ISSUE TITLE
Allow me to run extended tests on an existing cluster

### ISSUE THREAD

**Issue by kargakis** [openshift/origin#9319]
Most if not all of the time I just want to run a specific test. Building up and tearing down a cluster, a registry, and a router every time for a single test that may not even use the infra components is not optimal. I should be able to run the test I want  on an already exisitng cluster.

@stevekuznetsov @smarterclayton @deads2k

**Comment by stevekuznetsov** [comment]
@derekwaynecarr @mfojtik since you two were so supportive of Ginkgo -- how reasonable is it to do this with our current framework?

**Comment by deads2k** [comment]
There are two broad categories of extended tests:
 1. tests that only need a kubeconfig file.
  1. tests that mess with cluster-wide objects in a way that affects others
  1. tests that are project scoped.
 1. tests that need a whacky server setup.

If we can properly categorize the tests, the breaking apart server init from running the tests makes a lot of sense.  I spoke with @soltysh and @mfojtik about it while they were in town and @soltysh seemed interested in the idea.

One further wrinkle, we need to ensure cleanup 1.ii. style tests.

**Comment by deads2k** [comment]
> @derekwaynecarr @mfojtik since you two were so supportive of Ginkgo -- how reasonable is it to do this with our current framework?

Much as I may dislike ginkgo, I don't think it makes this problem easier or harder.  We have to decide what a canonical server looks like, then start scoping and tagging the buckets.

**Comment by mfojtik** [comment]
@deads2k i'm OK with tagging some tests to run against real cluster but we will have to be careful in cleaning things up. Also as @deads2k mentioned, we can only test things that does not require admin perms... 

@stevekuznetsov are we thinking about running a cluster in Jenkins setup? or is this more like "we want to share the cluster with other tests to speed up things" ?

**Comment by kargakis** [comment]
I was thinking of a flag I could use that would run the extended test or suite I want in my own cluster that I use for development. It would also be better for debugging.

**Comment by smarterclayton** [comment]
You mean... TEST_ONLY=1?

**Comment by smarterclayton** [comment]
or 

```
hack/build-go.sh test/extended/extended.test
extended.test --ginkgo.focus='suite'
```

?

**Comment by smarterclayton** [comment]
Or

```
oc cluster up --version=VERSION_I_WANT_TO_TEST
oc login -u system:admin
hack/build-go.sh test/extended/extended.test
extended.test --ginkgo.focus='suite'
```

**Comment by smarterclayton** [comment]
That already exists.  It's KUBECONFIG

(also, this was a humorous close, not a mean close.  I should have added an
emoji).

**Comment by smarterclayton** [comment]
Emoji added.

**Comment by kargakis** [comment]
> You mean... TEST_ONLY=1?

Ah, sweet

**Comment by deads2k** [comment]
> You mean... TEST_ONLY=1?

I think its reasonable to refactor the launch script to allow specifying a different cluster set-up mechanism to extended the tests.  In @soltysh's case, the goal is to run conformance tests with a particular config.  He doesn't want to do it as a one off, he wants to do it as an extended run on the AMIs.

**Comment by smarterclayton** [comment]
```
// run setup
TEST_ONLY=1 hack/test-extended.sh
```

**Comment by stevekuznetsov** [comment]
How does `set_up_the_cluster_like_i_want.sh` dynamically determine what I want?

**Comment by deads2k** [comment]
@stevekuznetsov I was thinking more like `test/extended/conformance.sh --config=test/extended/setup/[multimaster|default|audit-on].sh`

**Comment by smarterclayton** [comment]
Probably because you tell it what you want.  There is no setup step that is
required to be integrated into the test in that fashion.

**Comment by smarterclayton** [comment]
Note that if you want to create an "init" script that is run by a
particular suite, that makes sense.  But it doesn't make much sense to make
conformance parametrizable to setup when you can implement setup outside of
conformance.  Inversion of concerns.

**Comment by deads2k** [comment]
Forcing a setup "registration" into a particular location will make it easier to track all variation and write automated automation.  There are other spots to do it, but I think the alternatives end up looking like our vagrant-openshift versus openshift problems.  I'd really prefer to keep setup contained if at all possible.

**Comment by smarterclayton** [comment]
I'm totally ok with keeping setup constrained.  I just think you should do
setup before you call the test, not by passing a script into the test.

**Comment by derekwaynecarr** [comment]
I think the focus flag covered all that I would offer to this discussion ;-)

**Comment by soltysh** [comment]
I've already started working on changing our extended tests in such way that we'll be able to setup the server in non-standard way (see #9220). Generally I've reshuffled the scripts so that it's easy to setup the server, change the configs (big kudos to @deads2k for `openshift ex config patch`) but still run the entire conformance suite. I've talked with @stevekuznetsov we'll add those to our daily runs so that we can ensure everything is working as expected, but at the same time doesn't ruin our merge queue with lengthy tests.
